### DISCUSSION TITLE
How do i enter the auxiliary and normal keys together?  And keyboard press event.

### DISCUSSION THREAD

**Discussion by kimcoldwater** [FreeRDP/FreeRDP#7973] [Q&A] [answered]
hello. 
I'm customizing FreeRDP for use on the web.
I am implementing keyboard events using javascript events.

I want to press "shift" and enter capital letters.

first , 
` freerdp_input_send_keyboard_event(input, keyFlags | KBD_FLAGS_DOWN, scancode);`
Forward the shift keycode to the DOWN event

next,
Pass the character key code to send_keyboard. ( DOWN & UP ) 

finally,
`freerdp_input_send_keyboard_event(input, keyFlags | KBD_FLAGS_Release, scancode);`
Forward UP event

I succeeded in entering the auxiliary key and the normal key separately.

However, you could not enter uppercase characters using lowercase and shift.

Is there a value or method that needs to be specified separately?

And in javascript and window, there are keydown, keydown, and keypress events.

I couldn't find how to deliver the 'press' event in the code.

How should I deliver the press event? Or is there a function to deliver a press event?

Thank you for reading my question.

**Comment by akallabeth** [comment, marked as answer]
No.
You just send the scancodes in question.
e.g.
```
shift down
key 'a' down
key 'a'up
shift up
```
and you should get the capital letter 'A' (assuming you have the correct keyboard layout locally and remotely)

If the server supports unicode input (client has `FreeRDP_UnicodeInput` set and server has `INPUT_FLAG_UNICODE` in input capabilities set) then you can use `freerdp_input_send_unicode_keyboard_event` to eliminate the problem with keyboard layouts, but you still need to send the shift events

**Reply by kimcoldwater** [reply]
Thank you for your answer.
Is ' press ' event realize in FreeRDP?

**Reply by akallabeth** [reply]
you posted it yourself, `freerdp_input_send_keyboard_event(input, keyFlags | KBD_FLAGS_DOWN, scancode);` (sorry, it is named `KBD_FLAGS_DOWN`, but that is exactly that - a keypress)

Example with `freerdp_input_send_keyboard_event_ex`:
```
freerdp_input_send_keyboard_event_ex(input, TRUE, scancode);
freerdp_input_send_keyboard_event_ex(input, FALSE, scancode); 
```
will simulate a keypress (e.g. `KBD_FLAGS_DOWN` and `KBD_FLAGS_RELEASE`)

**Reply by kimcoldwater** [reply]
I understand that there are events called 'up', 'down' and 'press' in Windows keyboard events.

Recently, i have confirmed that it is not being used well.

But I'm want to make it for situations where [press] events are needed.

I understand the above code as 'up' and 'down' events.

how i forward the 'press' event?

**Reply by akallabeth** [reply]
there is no `press` event, there is just `down` and `up`.
A `press` is the sequence of `down` and `up` events
This is used in all of `FreeRDP` clients (and others utilizing `FreeRDP`) so you seem to have an issue with something else (maybe wrong scancode?)

**Reply by kimcoldwater** [reply]
ok.

I checked to the Apache guacamole sauce.

I asked because the press event was defined in Guacamole implemented as FreeRDP.

I don't know if I was mistaken because I didn't look closely.

Thank you for always answering with care.

It's be a great help to me.
